llvm/test/CodeGen/AMDGPU/llvm.fptrunc.round.ll

```diff
@@ -3,6 +3,15 @@
 ; RUN: llc -global-isel=0 -mtriple=amdgcn -mcpu=gfx1010 < %s | FileCheck -check-prefixes=CHECK,SDAG %s
 ; RUN: llc -global-isel -mtriple=amdgcn -mcpu=gfx1030 < %s | FileCheck -check-prefixes=CHECK,GISEL %s
 
+define amdgpu_gs half @v_fptrunc_round_f32_to_f16_tonearest(float %a) {
+; CHECK-LABEL: v_fptrunc_round_f32_to_f16_tonearest:
+; CHECK:       ; %bb.0:
+; CHECK-NEXT:    v_cvt_f16_f32_e32 v0, v0
+; CHECK-NEXT:    ; return to shader part epilog
+  %res = call half @llvm.fptrunc.round.f16.f32(float %a, metadata !"round.tonearest")
```

Comment:
this should work, but hopefully instcombine just turns this into fptrunc?

Reply:
At custom lowering the the intrinsic, we can turn it to fp_round, which can also generated the expected output for "round.tonearest". However, it seems the lowering of fp_round ignores the rounding mode operand in most cases (just assume default of round-to-nearest-even)

Reply:
that shouldn't be custom lowering, that should be a generic lowering (and instcombine optimization)

Reply:
I would put the instcombine optimization for llvm.fptrunc.round of mode round.tonearest to a separate patch. Thanks.

Reply:
Can we do something in instcombine for round.towardzero also?

Reply:
> I would put the instcombine optimization for llvm.fptrunc

Yes

> Can we do something in instcombine for round.towardzero also?

I assume quite a few combines could be done with it, but you can't merely replace it with anything obvious

Reply:
For f32 -> bf16, you won't be able to rtz for v_cvt_pk_bf16_f32 (hardware uses round to nearest even all the times).
So if you really want to rtz, you will have to do in the compiler.

Reply:
Yes, it will need some kind of expansion